pymatgen/analysis/phase_diagram.py

```diff
@@ -599,14 +599,98 @@ def get_equilibrium_reaction_energy(self, entry):
+
+    def get_decomposition_energy(self, entry):
+        """
+        NOTE a different name might be better to avoid confusion with
+        the decomposition of unstable compounds.
+
+        Provides the decomposition energy of an entry from the neighboring
+        stable entries excluding the entry in question. For unstable entries
+        this is the same as the energy above the hull.
+
+        Args:
+            entry: A PDEntry like object
+
+        Returns:
+            Decomposition energy per atom of entry. Stable entries should have
+            decomposition energies <= 0.
+        """
+        # Handle unstable materials
+        if entry not in self.stable_entries:
+            return self.get_e_above_hull(entry)
+
+        # Handle stable elemental materials
+        if entry.is_element:
+            return 0
+
+        # Handle stable compounds
+        return self.get_decomp_and_decomp_energy(entry)[1]
+
+    def get_decomp_and_decomp_energy(self, entry, space_limit=200):
+        """
+        Args:
+            entry (PDEntry): A PDEntry like object.
+            space_limit (int): The maximum number of competing entries to consider.
+
+        Returns:
+            (decomp, energy), decomp is given as a dict of {PDEntry, amt} for all
+            entries in the decomp reaction. On failure returns ({}, NaN) if stable.
+        """
+        # For unstable materials use simplex approach
+        if entry not in self.stable_entries:
+            return self.get_decomp_and_e_above_hull(entry)
+
+        # take entries with negative formation enthalpies as competing entries
+        competing_entries = [c for c in self.qhull_entries if c != entry
+                             if set(c.composition.elements).issubset(entry.composition.elements)]
+
+        # TODO could potentially avoid optimization in rare cases where the only competing
+        # entries are elements as then the solution is just the formation energy.
+
+        # NOTE SLSQP optimizer doesn't scale well for > 300 competing entries. As a
+        # result in phase diagrams where we have too many competing entries we can
+        # reduce the number by looking at the first and second convex hulls. This
+        # requires computing the convex hull of a second (hopefully smallish) space
+        # and so is not done by default
+        if len(competing_entries) > space_limit:
+            inner_hull = PhaseDiagram(list(set.intersection(
+                set(competing_entries),  # same chemical space
+                set(self.qhull_entries),  # negative E_f
+                set(self.unstable_entries),  # not already on hull
+            )) + list(self.el_refs.values()))  # terminal points
+
+            competing_entries = list(self.stable_entries.union(inner_hull.stable_entries))
+            competing_entries = [c for c in competing_entries if c != entry]
+
+        solution = _slsqp_decomp_solution(entry, competing_entries)
```

Comment:
If speed is an issue, I'm ok with adding `pyOpt` as an optional dependency (since it'd probably be useful elsewhere too), and adding a `try/except ImportError` in this module to fall-back to the SciPy version if necessary. Don't feel strongly on this though, so your call.